Update polyfill to use WebVR v1 API

toji opened the pull request:
Now includes distortion and querying of device parameters from DPDB. Old API
is still accessible when setting WebVRConfig.ENABLE_DEPRECATED_API to true.

toji commented:
Apologies for the large pull request. Happy to discuss any part of it you have questions about. Would like to try and get this merged before the release of the WebVR v1 spec/builds so that I can point developers to the official polyfill repo from day one instead of my fork.

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+        self.isPresenting = fullscreenElement == layer.source;
+        self.fireVRDisplayPresentChange_();
+        if (self.isPresenting) {
+          if (screen.orientation && screen.orientation.lock)
```
jshint will complain; I would put `/* global screen */` up top

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+          if (screen.orientation && screen.orientation.lock)
+            screen.orientation.lock("landscape-primary");
+          self.waitingForPresent_ = false;
+          self.beginPresent_();
```
wake about starting wakelock?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+          onFullscreenChange, onFullscreenError);
+
+      self.waitingForPresent_ = true;
+      // TODO: Spec says that requestFullscreen returns a promise, but that
```
yeah, the Promise behaviour described in the WHATWG Fullscreen spec hasn't shipped anywhere yet

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+      self.waitingForPresent_ = true;
+      // TODO: Spec says that requestFullscreen returns a promise, but that
+      // doesn't seem to be implemented anywhere just yet.
+      if (layer.source.requestFullscreen)
```
want to create a single generic function for entering FS?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+    }
+
+    if (!self.waitingForPresent_) {
+      document.exitFullscreen();
```
what about `mozCancelFullScreen`, `webkitExitFullscreen`, and `msExitFullscreen`?

can you just call `self.exitPresent()`?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+    if (wasPresenting) {
+      // TODO: Spec says that exitFullscreen returns a promise, but that
+      // doesn't seem to be implemented anywhere just yet.
+      if (document.exitFullscreen)
```
can we make this DRY too?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+
+VRDisplay.prototype.requestPresent = function(layer) {
+  var self = this;
+  this.layer_ = layer;
```
do we want to add a comment or support passing in an array of layers just yet?

toji replied:
It's not in the spec yet, so I don't think we need to complicate this code ATM.

cvan replied:
ok, that's fair. could we add [a comment here](https://github.com/borismus/webvr-polyfill/pull/33/files#diff-d7e697eff3913f1acaacac8002c3b05eR158) about using an array?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+    self.waitingForPresent_ = false;
+    if (layer && layer.source) {
+      function onFullscreenChange() {
+        var fullscreenElement = document.fullscreenElement ||
```
perhaps move this into `Util`?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+            document.mozFullScreenElement ||
+            document.msFullscreenElement;
+
+        self.isPresenting = fullscreenElement == layer.source;
```
`===`?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+  if (this.layer_) {
+    return [this.layer_];
+  }
+  return null;
```
what about returning an empty array instead of `null`?

toji replied:
Spec currently lists this function as returning `sequence<VRLayer>?`, I figured that since it's nullable that implies that `null` is the default return value when nothing is set. If that's not the intended behavior we should change the spec to make the return value non-nullable. Either way the spec should have text explaining what the expected return values are.

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+
+  if (this.fullscreenChangeHandler_) {
+    var changeHandler = this.fullscreenChangeHandler_;
+    element.removeEventListener("fullscreenchange", changeHandler, false);
```
perhaps it'd be helpful to have a helper method `removeFullscreenEventListener_` that would handle all the vendor-prefixed event names?

cvan commented on src/base.js:
```diff
@@ -13,8 +13,225 @@
+  if (this.fullscreenChangeHandler_) {
+    var changeHandler = this.fullscreenChangeHandler_;
+    element.removeEventListener("fullscreenchange", changeHandler, false);
+    element.removeEventListener("webkitfullscreenchange", changeHandler, false);
```
single quotes are used in a lot of other places

toji commented:
Addressed @cvan's comments and fixed a bug or two.

cvan commented on src/webvr-polyfill.js:
```diff
@@ -13,46 +13,81 @@
+  this.nativeWebVRAvailable = this.isWebVRAvailable();
 
-  if (!this.isWebVRAvailable()) {
+  if (!this.nativeWebVRAvailable) {
```
IMO, we should bail `return` early if `this.nativeWebVRAvailable` (would also prevent having to [check `WebVRConfig.ENABLE_DEPRECATED_API` below](#diff-7bfb91b61bb80cff41aba297525c4dc7R35) if you wanted to).

cvan commented on src/webvr-polyfill.js:
```diff
@@ -13,46 +13,81 @@
+  // TODO: For the moment this is only available via the deprecated APIs.
+  // Should refactor them to be exposed as VRDisplays that are wrapped as
+  // VRDevices.
+  if (WebVRConfig.ENABLE_DEPRECATED_API && !this.isMobile()) {
```
can we have `WebVRConfig.ENABLE_DEPRECATED_API` default to `true` for now?

cvan commented:
I had to make these changes for this to work in Firefox and your older Chromium builds:

```diff
diff --git a/src/webvr-polyfill.js b/src/webvr-polyfill.js
index 0e8c237..a02952b 100644
--- a/src/webvr-polyfill.js
+++ b/src/webvr-polyfill.js
@@ -28,12 +28,15 @@ function WebVRPolyfill() {
   this.devices = []; // For deprecated objects
   this.devicesPopulated = false;
   this.nativeWebVRAvailable = this.isWebVRAvailable();
+  this.nativeLegacyWebVRAvailable = this.isDeprecatedWebVRAvailable();
 
-  if (!this.nativeWebVRAvailable) {
-    this.enablePolyfill();
-  }
-  if (WebVRConfig.ENABLE_DEPRECATED_API && !this.isDeprecatedWebVRAvailable()) {
-    this.enableDeprecatedPolyfill();
+  if (!this.nativeLegacyWebVRAvailable) {
+    if (!this.nativeWebVRAvailable) {
+      this.enablePolyfill();
+    }
+    if (WebVRConfig.ENABLE_DEPRECATED_API) {
+      this.enableDeprecatedPolyfill();
+    }
   }
 }
 
@@ -113,15 +116,30 @@ WebVRPolyfill.prototype.getVRDevices = function() {
   var self = this;
   return new Promise(function(resolve, reject) {
     try {
-      if (!self.devicesPopulated && self.nativeWebVRAvailable) {
-        navigator.getVRDisplays(function(displays) {
-          for (var i = 0; i < displays.length; ++i) {
-            self.devices.push(new VRDisplayHMDDevice(displays[i]));
-            self.devices.push(new VRDisplayPositionSensorDevice(displays[i]));
-          }
-          self.devicesPopulated = true;
-          resolve(self.devices);
-        }, reject);
+      if (!self.devicesPopulated) {
+        if (self.nativeWebVRAvailable) {
+          navigator.getVRDisplays(function(displays) {
+            for (var i = 0; i < displays.length; ++i) {
+              self.devices.push(new VRDisplayHMDDevice(displays[i]));
+              self.devices.push(new VRDisplayPositionSensorDevice(displays[i]));
+            }
+            self.devicesPopulated = true;
+            resolve(self.devices);
+          }, reject);
+        } else if (self.nativeLegacyWebVRAvailable) {
+          (navigator.getVRDDevices || navigator.mozGetVRDevices)(function(devices) {
+            for (var i = 0; i < devices.length; ++i) {
+              if (devices[i] instanceof HMDVRDevice) {
+                self.devices.push(displays[i]);
+              }
+              if (devices[i] instanceof PositionSensorVRDevice) {
+                self.devices.push(devices[i]);
+              }
+            }
+            self.devicesPopulated = true;
+            resolve(self.devices);
+          }, reject);
+        }
       } else {
         self.populateDevices();
         resolve(self.devices);
```

cvan commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+
```
I'd recommend

```
*.log
*.pid
*.seed
*~
.DS_Store
.node_repl_history
.npm
logs
node_modules
npm-debug.log*
pids
```

cvan replied:
I opened #34; this is fine for now

cvan commented:
latest patch to fix Firefox/older Chromium experimental builds:

```diff
diff --git a/src/webvr-polyfill.js b/src/webvr-polyfill.js
index 0e8c237..b2c12f9 100644
--- a/src/webvr-polyfill.js
+++ b/src/webvr-polyfill.js
@@ -28,12 +28,15 @@ function WebVRPolyfill() {
   this.devices = []; // For deprecated objects
   this.devicesPopulated = false;
   this.nativeWebVRAvailable = this.isWebVRAvailable();
+  this.nativeLegacyWebVRAvailable = this.isDeprecatedWebVRAvailable();
 
-  if (!this.nativeWebVRAvailable) {
-    this.enablePolyfill();
-  }
-  if (WebVRConfig.ENABLE_DEPRECATED_API && !this.isDeprecatedWebVRAvailable()) {
-    this.enableDeprecatedPolyfill();
+  if (!this.nativeLegacyWebVRAvailable) {
+    if (!this.nativeWebVRAvailable) {
+      this.enablePolyfill();
+    }
+    if (WebVRConfig.ENABLE_DEPRECATED_API) {
+      this.enableDeprecatedPolyfill();
+    }
   }
 }
 
@@ -113,21 +116,36 @@ WebVRPolyfill.prototype.getVRDevices = function() {
   var self = this;
   return new Promise(function(resolve, reject) {
     try {
-      if (!self.devicesPopulated && self.nativeWebVRAvailable) {
-        navigator.getVRDisplays(function(displays) {
-          for (var i = 0; i < displays.length; ++i) {
-            self.devices.push(new VRDisplayHMDDevice(displays[i]));
-            self.devices.push(new VRDisplayPositionSensorDevice(displays[i]));
-          }
-          self.devicesPopulated = true;
+      if (!self.devicesPopulated) {
+        if (self.nativeWebVRAvailable) {
+          navigator.getVRDisplays(function(displays) {
+            for (var i = 0; i < displays.length; ++i) {
+              self.devices.push(new VRDisplayHMDDevice(displays[i]));
+              self.devices.push(new VRDisplayPositionSensorDevice(displays[i]));
+            }
+            self.devicesPopulated = true;
+            resolve(self.devices);
+          }, reject);
+        } else if (self.nativeLegacyWebVRAvailable) {
+          (navigator.getVRDDevices || navigator.mozGetVRDevices)(function(devices) {
+            for (var i = 0; i < devices.length; ++i) {
+              if (devices[i] instanceof HMDVRDevice) {
+                self.devices.push(displays[i]);
+              }
+              if (devices[i] instanceof PositionSensorVRDevice) {
+                self.devices.push(devices[i]);
+              }
+            }
+            self.devicesPopulated = true;
+            resolve(self.devices);
+          }, reject);
+        } else {
+          self.populateDevices();
           resolve(self.devices);
-        }, reject);
-      } else {
-        self.populateDevices();
-        resolve(self.devices);
+        }
+      } catch (e) {
+        reject(e);
       }
-    } catch (e) {
-      reject(e);
     }
   });
 };
```

cvan commented on src/webvr-polyfill.js:
```diff
@@ -13,46 +13,81 @@
   // Initialize our virtual VR devices.
   if (this.isCardboardCompatible()) {
-    this.devices.push(new CardboardHMDVRDevice());
+    var vrDisplay = new CardboardVRDisplay();
```
variable already defined [above](#diff-7bfb91b61bb80cff41aba297525c4dc7R52). drop the `var`.

toji commented:
Implemented the patch @cvan pasted inline above as well as a couple of other fixes mentioned in this conversation. Also refactored the mouse and keyboard implementation as a VRDisplay and started exposing it on non-mobile. As always, you can see the latest version in action on http://toji.github.io/webvr-samples/03a-webvr-polyfill.html

toji commented:
Latest update fixes resetPose on CardboardVRDisplay and corrects non-black distortion background with backbuffers that have an alpha channel. (which is forced on iOS)

envokevr commented on src/base.js:
```diff
@@ -13,8 +13,228 @@
+};
+
+VRDisplay.prototype.fireVRDisplayPresentChange_ = function() {
+  var event = new CustomEvent('vrdisplaypresentchange', { 'vrdisplay': this });
```
@toji listeners to this event don't receive the `vrdisplay` value, it needs to be added to the event detail:

```js
var event = new CustomEvent('vrdisplaypresentchange', { detail: { vrdisplay: this } });
```

which makes it available at `e.details.vrdisplay`

Deminetix replied:
@toji listeners to this event don't receive the `vrdisplay` value, it needs to be added to the event detail:

```js
var event = new CustomEvent('vrdisplaypresentchange', { detail: { vrdisplay: this } });
```

which makes it available at `e.details.vrdisplay`

cvan replied:
I was going to reply here, but I instead adding a comment to the issue filed in the spec repo. Feel free to drop a line there: MozVR/webvr-spec#15.

toji commented:
Added flag to defer initialization of the polyfill (now used with all the samples at https://toji.github.io/webvr-samples) and fixed the issue @ashconnell brought up.

cvan commented on README.md:
```diff
@@ -54,5 +54,10 @@ are supported:
+      //MOUSE_KEYBOARD_CONTROLS_DISABLED: true, // Default: false
+      // Prevent the polyfill from initializing immediately. Requiures the app
+      // to call InitializeWebVRPolyfill() before it can be used.
+      //DEFER_INITIALIZATION: true, // default: false
```
would also mention this in the commit message

cvan commented on src/main.js:
```diff
@@ -16,4 +16,11 @@ var WebVRPolyfill = require('./webvr-polyfill.js');
-new WebVRPolyfill();
+
+if (!window.WebVRConfig.DEFER_INITIALIZATION) {
+  new WebVRPolyfill();
```
for CommonJS/AMD, I would recommend changing to this:

```js
module.exports = new WebVRPolyfill();
```

cvan replied:
in fact, I would recommend opening a separate PR completely for this defer initialisation behaviour. I'll _defer_ to @borismus obviously as this is his library. but I think it'd make sense to just make breaking changes.

I'd suggest (à la [the `es6-promise` polyfill](https://github.com/stefanpenner/es6-promise#auto-polyfill)) to require the user to explicitly call a method on an exported module to polyfill.

so instead of having the `WebVRPolyfill` constructor automatically initialise the [polyfill](https://github.com/borismus/webvr-polyfill/blob/ca30e6d/src/webvr-polyfill.js#L28-L30),

so `WebVRPolyfill` would be exported and [these bits would be removed from the constructor](https://github.com/borismus/webvr-polyfill/blob/ca30e6d/src/webvr-polyfill.js#L28-L30) and [this line](https://github.com/borismus/webvr-polyfill/pull/33/files#diff-1c90ff38a08209f9ebd4d05d1e43358eL19) would get replaced with `module.exports = WebVRPolyfill` (and browserify would then attach that to `window`, handle AMD, CommonJS, etc.).

cvan commented on src/main.js:
```diff
@@ -16,4 +16,11 @@ var WebVRPolyfill = require('./webvr-polyfill.js');
+if (!window.WebVRConfig.DEFER_INITIALIZATION) {
+  new WebVRPolyfill();
+} else {
+  window.InitializeWebVRPolyfill = function() {
```
for CommonJS/AMD, I would recommend adding this:

```js
module.exports.polyfill = function () {
  return new WebVRPolyfill();
};
```

cvan replied:
[kind of like this](https://github.com/stefanpenner/es6-promise#auto-polyfill)

cvan commented:
`build/webvr-polyfill.js` needs to be rebuilt. or I guess you could undo those changes and wait until the PR is ready to be merged and @borismus could do it.

toji commented:
Latests patch now includes basic Cardboard UI with the alignment line and options button. Hooks are in for a button click listener but at the moment it only writes a "not implemented" error to the console.

![screenshot_20160302-102041](https://cloud.githubusercontent.com/assets/805273/13470584/bd4a759a-e061-11e5-82e9-fd27a0266dc5.png)

Also made sure I rebuilt the `build/` version this time. :)

As for the deferred load thing, I talked to @borismus about it offline yesterday and he seemed OK with the current form, but if he'd like me to refactor it I'm happy to.

borismus commented:
Thanks Brandon, this is huge. I've started a parallel webvr-boilerplate branch. Once everything is working (hopefully with compat), can branch both branches to masters.

cvan commented:
woo! thanks, @borismus and @toji! :+1:
